src/reconcile/rec_write.c

```diff
@@ -529,11 +530,13 @@ __rec_init(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags, WT_SALVAGE_COO
      * can move beyond the checkpoint transaction id. When reconciling the metadata, we have to take
      * checkpoints into account.
      */
-    if (WT_IS_METADATA(session->dhandle)) {
+    if (WT_IS_METADATA(session->dhandle))
         WT_ORDERED_READ(ckpt_txn, txn_global->checkpoint_txn_shared.id);
-        if (ckpt_txn != WT_TXN_NONE && WT_TXNID_LT(ckpt_txn, r->last_running))
-            r->last_running = ckpt_txn;
-    }
+    else if (!F_ISSET(r, WT_REC_CHECKPOINT))
+        WT_ORDERED_READ(ckpt_txn, txn_global->checkpoint_txn_shared.pinned_id);
```

Comment:
This is the other portion of the change to handle when eviction gets ahead of checkpoint.

In this case, we're going to ensure that if a checkpoint is already running, our `r->last_running` won't go past the checkpoint's pinned timestamp. So our eviction mimick the checkpoint's visibility.

As @quchenhao pointed out, this is only an issue when we have failed evictions. This is normally not an issue because eviction will free updates that got written to the history store or the page so we won't have this problem of checkpoint choosing older updates. However, if we fail an eviction, these updates may be written to the history store and still exist in the chain for the checkpoint to see.